File: client/java/src/main/java/com/gomelo/ProtobufCodec.java
package com.gomelo;

import com.google.protobuf.Message;
import com.google.protobuf.Parser;

import java.nio.ByteBuffer;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Supplier;

public class ProtobufCodec {

    private final Map<String, Integer> routeToId = new HashMap<>();
    private final Map<Integer, String> idToRoute = new HashMap<>();
    private final Map<String, Parser<? extends Message>> parsers = new HashMap<>();
    private int nextId = 0;

    public void registerRoute(String route, int id) {
        routeToId.put(route, id);
        idToRoute.put(id, route);
    }

    public void registerType(String route, int id, Parser<? extends Message> parser) {
        registerRoute(route, id);
        parsers.put(route, parser);
    }

    public int getRouteId(String route) {
        return routeToId.getOrDefault(route, 0);
    }

    public String getRoute(int id) {
        return idToRoute.getOrDefault(id, "");
    }

    public byte[] encode(int msgType, String route, long seq, Message body) {
        int routeId = routeToId.getOrDefault(route, 0);
        byte[] bodyBytes = body != null ? body.toByteArray() : new byte[0];

        int headerLen;
        if (routeId > 0) {
            headerLen = 1 + 2 + 8;
        } else {
            headerLen = 1 + 1 + route.length() + 1 + 8;
        }

        byte[] result = new byte[4 + headerLen + bodyBytes.length];

        int totalLen = headerLen + bodyBytes.length;
        result[0] = (byte) ((totalLen >> 24) & 0xFF);
        result[1] = (byte) ((totalLen >> 16) & 0xFF);
        result[2] = (byte) ((totalLen >> 8) & 0xFF);
        result[3] = (byte) (totalLen & 0xFF);

        int offset = 4;
        result[offset++] = (byte) msgType;

        if (routeId > 0) {
            result[offset++] = 0x01;
            result[offset++] = (byte) ((routeId >> 8) & 0xFF);
            result[offset++] = (byte) (routeId & 0xFF);
        } else {
            result[offset++] = 0x00;
            for (int i = 0; i < route.length(); i++) {
                result[offset++] = (byte) route.charAt(i);
            }
            result[offset++] = 0;
        }

        result[offset++] = (byte) ((seq >> 56) & 0xFF);
        result[offset++] = (byte) ((seq >> 48) & 0xFF);
        result[offset++] = (byte) ((seq >> 40) & 0xFF);
        result[offset++] = (byte) ((seq >> 32) & 0xFF);
        result[offset++] = (byte) ((seq >> 24) & 0xFF);
        result[offset++] = (byte) ((seq >> 16) & 0xFF);
        result[offset++] = (byte) ((seq >> 8) & 0xFF);
        result[offset++] = (byte) (seq & 0xFF);

        System.arraycopy(bodyBytes, 0, result, offset, bodyBytes.length);

        return result;
    }

    public DecodedMessage decode(byte[] data, int offset, int length) {
        if (length < 5) {
            return null;
        }

        int totalLen = ((data[offset] & 0xFF) << 24) | ((data[offset + 1] & 0xFF) << 16) |
                       ((data[offset + 2] & 0xFF) << 8) | (data[offset + 3] & 0xFF);

        if (totalLen > 64 * 1024 || totalLen == 0 || totalLen + 4 > length) {
            return null;
        }

        int msgType = data[offset + 4];
        int pos = offset + 5;

        String route = null;
        int responseSeq = 0;

        if (msgType == GomeloClient.MessageType.Response.getValue()) {
            if (pos + 4 > offset + length) return null;
            responseSeq = ((data[pos] & 0xFF) << 24) | ((data[pos + 1] & 0xFF) << 16) |
                         ((data[pos + 2] & 0xFF) << 8) | (data[pos + 3] & 0xFF);
            pos += 4;
        } else {
            if (pos >= offset + length) return null;
            if (data[pos] == 0x01) {
                pos++;
                if (pos + 2 > offset + length) return null;
                int routeId = (data[pos] << 8) | data[pos + 1];
                pos += 2;
                route = getRoute(routeId);
            } else if (data[pos] == 0x00) {
                pos++;
                int start = pos;
                while (pos < offset + length && data[pos] != 0) pos++;
                route = new String(data, start, pos - start);
                pos++;
            }
        }

        if (pos > offset + length) return null;

        int bodyLen = offset + length - pos;
        byte[] bodyBytes = new byte[bodyLen];
        System.arraycopy(data, pos, bodyBytes, 0, bodyLen);

        Message bodyMsg = null;
        if (route != null && parsers.containsKey(route) && bodyBytes.length > 0) {
            try {
                @SuppressWarnings("unchecked")
                Parser<Message> parser = (Parser<Message>) parsers.get(route);
                bodyMsg = parser.parseFrom(bodyBytes);
            } catch (Exception e) {
                // fallback to raw bytes
            }
        }

        return new DecodedMessage(msgType, route, responseSeq, bodyMsg, bodyBytes);
    }

    public static class DecodedMessage {
        public final int type;
        public final String route;
        public final long seq;
        public final Message body;
        public final byte[] rawBody;

        public DecodedMessage(int type, String route, long seq, Message body, byte[] rawBody) {
            this.type = type;
            this.route = route;
            this.seq = seq;
            this.body = body;
            this.rawBody = rawBody;
        }
    }
}

File: client/js/protobuf_codec.js
/**
 * Gomelo Protobuf Codec for JavaScript Client
 * Handles message framing only - body bytes are passed through as-is
 * Server is responsible for PB serialization/deserialization
 */

const MessageType = {
  Request: 1,
  Response: 2,
  Notify: 3,
  Error: 4
};

const RouteFlag = {
  RouteID: 0x01,
  RouteString: 0x00
};

class ProtobufCodec {
  constructor() {
    this.routeToId = new Map();
    this.idToRoute = new Map();
    this.nextId = 0;
  }

  registerRoute(route, id) {
    this.routeToId.set(route, id);
    this.idToRoute.set(id, route);
  }

  registerMessage(route, id) {
    this.registerRoute(route, id);
  }

  getRouteId(route) {
    return this.routeToId.get(route) || 0;
  }

  getRoute(id) {
    return this.idToRoute.get(id) || '';
  }

  encode(msgType, route, seq, bodyBytes) {
    const routeId = this.getRouteId(route);
    const body = bodyBytes || Buffer.alloc(0);

    let headerLen;
    let routePart;

    if (routeId > 0) {
      headerLen = 1 + 1 + 2 + 8;
      routePart = Buffer.alloc(3);
      routePart[0] = RouteFlag.RouteID;
      routePart.writeUInt16BE(routeId, 1);
    } else {
      const routeBytes = Buffer.from(route, 'utf8');
      headerLen = 1 + 1 + routeBytes.length + 1 + 8;
      routePart = Buffer.alloc(routeBytes.length + 2);
      routePart[0] = RouteFlag.RouteString;
      routeBytes.copy(routePart, 1);
      routePart[routePart.length - 1] = 0;
    }

    const buf = Buffer.alloc(1 + headerLen + body.length);
    let offset = 0;

    buf[offset] = msgType;
    offset += 1;

    routePart.copy(buf, offset);
    offset += routePart.length;

    buf.writeBigUInt64BE(BigInt(seq), offset);
    offset += 8;

    body.copy(buf, offset);

    return buf;
  }

  decode(data) {
    const buf = Buffer.isBuffer(data) ? data : Buffer.from(data);
    if (buf.length < 10) {
      return null;
    }

    const msgType = buf[0];
    let offset = 1;

    const flag = buf[offset];
    offset += 1;

    let route = '';
    let routeId = 0;
    if (flag === RouteFlag.RouteID) {
      routeId = buf.readUInt16BE(offset);
      offset += 2;
      route = this.idToRoute.get(routeId) || '';
    } else if (flag === RouteFlag.RouteString) {
      const start = offset;
      while (offset < buf.length && buf[offset] !== 0) {
        offset++;
      }
      route = buf.toString('utf8', start, offset);
      offset++;
    }

    if (offset + 8 > buf.length) {
      return null;
    }

    const seq = Number(buf.readBigUInt64BE(offset));
    offset += 8;

    const bodyBytes = offset < buf.length ? buf.slice(offset) : Buffer.alloc(0);

    let body = null;
    if (bodyBytes.length > 0) {
      try {
        body = JSON.parse(bodyBytes.toString());
      } catch {
        body = bodyBytes;
      }
    }

    return {
      type: msgType,
      route: route,
      routeId: routeId,
      seq: seq,
      body: body
    };
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { ProtobufCodec, MessageType, RouteFlag };
}

if (typeof window !== 'undefined') {
  window.ProtobufCodec = ProtobufCodec;
  window.MessageType = MessageType;
  window.RouteFlag = RouteFlag;
}
